cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -311,8 +310,7 @@ private void doForAllAppsWithCapability(String capability, CheckedConsumer<Appli
 
   private boolean shouldRetry(Throwable throwable) {
     return !(throwable instanceof UnauthorizedException ||
-      throwable instanceof InvalidArtifactException ||
-      throwable instanceof ArtifactNotFoundException);
```

Comment:
What if the artifact is deleted and gone? Should we retry forever?

Reply:
Seems like there is a retry limit, please ignore this comment

Reply:
There is a retry limit which is set currently at 5.